doc/developers/contributing.rst

```diff
@@ -1143,6 +1143,12 @@ data dependent. A tolerance stopping criterion ``tol`` is not directly
 data dependent (although the optimal value according to some scoring
 function probably is).
 
+When ``fit`` is called, any previous call to ``fit`` should be ignored. In
+general, calling ``estimator.fit(X1)`` and then ``estimator.fit(X2)`` should
+be the same as only calling ``estimator.fit(X2)``. However, this may not be
+true in practice when ``fit`` depends on some random process, see
+:term:`random_state`.
+
```

Comment:
A second exception to this rule is when the hyper-parameter `warm_start` is set to `True` for estimators that support it. `warm_start=True` means that the previous state of the trainable parameters of the estimator are reused instead of using the default initialization strategy. If the optimization problem that backs the estimator is not guaranteed to admit a unique solution, the state of `estimator` after ``estimator.fit(X2)` might depend on a past call  `estimator.fit(X1)``.

Reply:
Good point!

Maybe we can keep it simple here and just mention that an other exception is `warm_start=True`? I guess the details could be left for the actual estimators docs.